File: rgctl-tests/ecommerce-typescript/src/coolstore/services/coolstoreOrderService.ts
import { createCoolstoreOrder, type CoolstoreOrder } from '../models/coolstoreOrder';
import { createCoolstoreOrderItem } from '../models/coolstoreOrderItem';
import type { ShoppingCart } from '../models/shoppingCart';

export class CoolstoreOrderService {
  private seq = 1;
  private readonly orders = new Map<number, CoolstoreOrder>();

  process(cart: ShoppingCart): CoolstoreOrder {
    const order = createCoolstoreOrder();
    order.orderId = this.seq++;
    order.cartId = cart.cartId;
    order.cartTotal = cart.cartTotal;
    order.items = [];
    for (const sci of cart.shoppingCartItemList) {
      if (sci.product) {
        order.items.push(
          createCoolstoreOrderItem(sci.product.itemId, sci.quantity, sci.price),
        );
      }
    }
    this.orders.set(order.orderId, order);
    return order;
  }

  getOrders(): CoolstoreOrder[] {
    return Array.from(this.orders.values());
  }

  getOrderById(orderId: number): CoolstoreOrder | undefined {
    return this.orders.get(orderId);
  }
}

File: rgctl-tests/ecommerce-csharp/src/Ecommerce/Coolstore/Services/CoolstoreOrderService.cs
using System.Collections.Concurrent;
using Ecommerce.Coolstore.Models;

namespace Ecommerce.Coolstore.Services;

public class CoolstoreOrderService
{
    private long _seq = 1;
    private readonly ConcurrentDictionary<long, CoolstoreOrder> _orders = new();

    public CoolstoreOrder Process(ShoppingCart cart)
    {
        var order = new CoolstoreOrder
        {
            OrderId = Interlocked.Increment(ref _seq) - 1,
            CartId = cart.CartId,
            CartTotal = cart.CartTotal,
            Items = new List<CoolstoreOrderItem>()
        };

        foreach (var sci in cart.ShoppingCartItemList)
        {
            if (sci.Product != null)
            {
                order.Items.Add(new CoolstoreOrderItem(
                    sci.Product.ItemId,
                    sci.Quantity,
                    sci.Price));
            }
        }

        _orders[order.OrderId] = order;
        return order;
    }

    public List<CoolstoreOrder> GetOrders() => _orders.Values.ToList();

    public CoolstoreOrder? GetOrderById(long orderId) =>
        _orders.TryGetValue(orderId, out var order) ? order : null;
}
